### PR TITLE
feat: hostels schema

### PR DESCRIPTION
**Pull request by heydoyouknowme0**
image
rest data improbable to change so left in i18n

### CODE REVIEW

**Comment by GetPsyched** [comment]
Wouldn't you want to initialise the hostels schema into master first, and then rebase `sample-db` onto `master`?

File: server/db/schema/hostels.schema.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  type: varchar('type', {
+    enum: ['boys', 'girls'],
+  }).notNull(),
+  createdAt: timestamp('created_at').defaultNow().notNull(),
+  updatedAt: timestamp('updated_at')
+    .$onUpdate(() => new Date())
+    .notNull(),
+});
+
+export const hostelRelations = relations(hostels, ({ many }) => ({
```
Comment (GetPsyched) [inline server/db/schema/hostels.schema.ts:22]: Minor nit: `Relations` should just be a suffix, meaning the rest should be the exact name of the table.
```suggestion
export const hostelsRelations = relations(hostels, ({ many }) => ({
```

File: server/db/schema/hostels.schema.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  posts: many(hostelPosts),
+}));
+
+export const hostelPosts = pgTable('hostel_posts', {
```
Comment (GetPsyched) [inline server/db/schema/hostels.schema.ts:26]: Reading that table name I thought it's for any articles that the hostel is releasing.
```suggestion
export const hostelStaff = pgTable('hostel_posts', {
```
Also, hostels have superintendents as well.

Probably look at the board outside the MMCA office or ask Praveen sir directly, he's chill

File: server/db/schema/hostels.schema.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  hostelId: serial('hostel_id')
+    .references(() => hostels.id)
+    .notNull(),
+  staffId: serial('staff_id')
+    .references(() => staff.id)
+    .notNull(),
+  post: varchar('post', { enum: ['warden', 'deputy-warden'] }).notNull(),
```
Comment (GetPsyched) [inline server/db/schema/hostels.schema.ts:33]: IDK about deputy wardens, but wardens are not staff, they're faculty.

**Review by GetPsyched — CHANGES_REQUESTED** [review]

File: server/db/schema/hostels.schema.ts (hunk 3 above)
Comment (heydoyouknowme0) [reply server/db/schema/hostels.schema.ts:33]: so should i make a single relation for staff and faculty or two different ones, cause practically same columns need to be retrieved in either

Comment (GetPsyched) [reply server/db/schema/hostels.schema.ts:33]: Given that only the warden/deputy-warden is a faculty member, I think it's best to offload that post to 2 columns in the `hostels` table (`wardenId`, and `deputyWardenId` I guess).

Then, leave this table as is, updating the enum to mention other values.

**Review by GetPsyched — CHANGES_REQUESTED** [review]

File: server/db/schema/hostels.schema.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  createdAt: timestamp('created_at').defaultNow().notNull(),
+  updatedAt: timestamp('updated_at')
+    .$onUpdate(() => new Date())
+    .notNull(),
+});
+
+export const hostelsRelations = relations(hostels, ({ many }) => ({
+  staff: many(hostelStaff),
+  wardens: many(faculty, { relationName: 'wardens' }),
+  deputyWardens: many(faculty, { relationName: 'deputy-wardens' }),
```
[the remark this answers is not in the document]
Comment (heydoyouknowme0) [reply server/db/schema/hostels.schema.ts:26]: there are 1-4 wardens depending upon the hostel

[the remark this answers is not in the document]
Comment (GetPsyched) [reply server/db/schema/hostels.schema.ts:26]: Oh, hmm.

Comment (GetPsyched) [inline server/db/schema/hostels.schema.ts:26]: If you have a `many` relationship, you need to have a `one` relationship somewhere else. See `doctorates` in `faculty` and `students` for example.

File: server/db/schema/hostels.schema.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+}));
+
+export const hostelStaff = pgTable('hostel_posts', {
+  hostelId: serial('hostel_id')
+    .references(() => hostels.id)
+    .notNull(),
+  staffId: serial('staff_id')
+    .references(() => staff.id)
+    .notNull(),
+  post: varchar('post', { enum: ['warden', 'deputy-warden'] }).notNull(),
```
Comment (GetPsyched) [inline server/db/schema/hostels.schema.ts:36]: Post won't be `'warden' | 'deputy-warden'` anymore. Please see other staff roles and put those here

**Review by GetPsyched — CHANGES_REQUESTED** [review]

File: server/db/schema/hostels.schema.ts (hunk 4 above)
Comment (heydoyouknowme0) [reply server/db/schema/hostels.schema.ts:26]: @KambojRajan suggested that it's not required.

Comment (GetPsyched) [reply server/db/schema/hostels.schema.ts:26]: Did you see the usage? Are you able to fetch a hsotel with its wardens appropriately?

Comment (heydoyouknowme0) [reply server/db/schema/hostels.schema.ts]: Sorry Im facing some problems with it. this was working as it is before now. It doesnt even work with one relation in faculty, but somehow if i reference different fields it starts working.

File: server/db/schema/hostels.schema.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    .notNull(),
+});
+
+export const hostelsRelations = relations(hostels, ({ many }) => ({
+  hostelStaff: many(hostelStaff),
+  wardens: many(faculty, { relationName: 'wardens' }),
+  deputyWardens: many(faculty, { relationName: 'deputy_wardens' }),
+}));
+
+export const hostelStaff = pgTable('hostel_posts', {
```
Comment (GetPsyched) [inline server/db/schema/hostels.schema.ts:29]:
```suggestion
export const hostelStaff = pgTable('hostel_staff', {
```

**Review by GetPsyched — CHANGES_REQUESTED** [review]

**Review by GetPsyched — APPROVED** [review]

**Comment by GetPsyched** [comment]
Wait, no. Put the overview and stuff into the schema. The data being improbable to change is no reason to not put it in there. Please don't use that as a criterion.

**Review by GetPsyched — APPROVED** [review]

**Comment by GetPsyched** [comment]
Rebase `sample-db` on `master` and add sample data in that branch

**Comment by heydoyouknowme0** [comment]
I messed up.
image
didnt realize employeeId is a varchar, it links properly inside studio but doesn't actually fetch when I try to fetch using the relation(which i realized when rebasing hostels page).

two many to one relation between same table dont work unless I use two different ids to do so

**Comment by GetPsyched** [comment]
As a lesson, always test your changes before moving the PR out of a draft

**Comment by heydoyouknowme0** [comment]
There was no indication of an error anywhere even after pushing to the db ˙◠˙ 

also what should I do now 
we can either make employee id an integer or another multi table relation(like staff)

**Comment by GetPsyched** [comment]
Um, that relation looks wrong though. Why is an employeeId being linked to a hostel id? Even if we made the datatypes match, why would they logically ever match?

**Comment by heydoyouknowme0** [comment]
uhhh, Im sorry everything's broken

**Pull request merged**